integreat_cms/cms/utils/repair_tree.py

```diff
@@ -0,0 +1,259 @@
+"""
+This module contains utilities to repair or detect inconsistencies in a tree
+"""
+from __future__ import annotations
+from django.db import transaction
+
+import logging
+
+from ..models import Page
+from ..utils.tree_mutex import tree_mutex
+
+logger = logging.getLogger(__name__)
+
+class Printer:
+    """
+    Select printer for stdout or log file
+    """
+    def __init__(self, print_func=None, error=None, success=None):
+        self._print = print_func
+        self._error = error
+        self._success = success
+        self._write = None
+        self._bold = None
+
+    @property
+    def print(self):
+        if not self._print:
+            return logger.debug
+        return self._print
+    @print.setter
+    def print(self, new):
+        self._print = new
+
+    @property
+    def error(self):
+        if not self._error:
+            return self.print
+        return self._error
+    @error.setter
+    def error(self, new):
+        self._error = new
+
+    @property
+    def success(self):
+        if not self._success:
+            return self.print
+        return self._success
+    @success.setter
+    def success(self, new):
+        self._success = new
+
+    @property
+    def bold(self):
+        if not self._bold:
+            return lambda x: x
+        return self._bold
+    @bold.setter
+    def bold(self, new):
+        self._bold = new
+
+    @property
+    def write(self):
+        if not self._write:
+            return self.print
+        return self._write
+    @write.setter
+    def write(self, new):
+        self._write = new
+
+
+@transaction.atomic
+@tree_mutex
+def repair_tree(page_id: int = 0, commit: bool = False, printer: Printer = Printer()) -> None:
+    pages_seen: list[int] = []
+    orphans = set()
+
+    mptt_fixer = MPTTFixer()
+
+    if page_id:
+        try:
+            page = Page.objects.get(id=page_id)
+        except Page.DoesNotExist as e:
+            raise ValueError(
+                f'The page with id "{page_id}" does not exist.'
+            ) from e
+        root_node = None
+        for tree_node in mptt_fixer.get_fixed_tree_of_page(page.pk):
+            # the first node is always the root node
+            if root_node is None:
+                root_node = tree_node
+            action = "Fixing" if commit else "Detecting problems in"
+            printer.print(f"{action} tree with id {root_node.tree_id}...")
+            print_changed_fields(Page.objects.get(id=tree_node.pk), tree_node.lft, tree_node.rgt, printer)
+        check_for_orphans(root_node.tree_id, pages_seen, printer)
```

Comment:
Record valid visited page as seen so it doesn't reported as orphan
```suggestion
            pages_seen.append(tree_node.pk)
        check_for_orphans(root_node.tree_id, pages_seen, printer)
```